core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -939,25 +939,33 @@ class KafkaController(val config: KafkaConfig,
 
   // Rearrange partition and replica assignment for new topics that get assigned to
   // maintenance brokers that do not take new partitions
-  private def rearrangePartitionReplicaAssignmentForNewTopics(topics: Set[String]) {
+  private def rearrangePartitionReplicaAssignmentForNewPartitions(topics: Set[String], onlyNewTopics: Boolean) {
     try {
       val noNewPartitionBrokers = partitionUnassignableBrokerIds
       if (noNewPartitionBrokers.nonEmpty) {
-        val newTopics = zkClient.getPartitionNodeNonExistsTopics(topics.toSet)
-        val newTopicsToBeArranged = zkClient.getPartitionAssignmentForTopics(newTopics).filter {
-          case (_, partitionMap) =>
-            partitionMap.exists {
+        val topicsToCheck = if (onlyNewTopics)
+          zkClient.getPartitionNodeNonExistsTopics(topics.toSet)
+        else
+          topics
+
+        val topicsToBeRearranged = zkClient.getPartitionAssignmentForTopics(topicsToCheck.toSet).filter {
+          case (topic, partitionMap) =>
+            val existingAssignment = controllerContext.partitionAssignments.getOrElse(topic, mutable.Map.empty)
+            val newPartitions = partitionMap.filter{case (partitionId, _) => partitionId >= existingAssignment.size}
```

Comment:
could we add some safe check here to ensure the partition state znode doesn't exist for these new partitions.  Although I don't see an issue in the current implementation, if this function is used incorrectly, it would be very dangerous.  For example,  if we use "rearrangePartitionReplicaAssignmentForNewPartitions(topics, false)"  when initializing controller context, this may cause all topics to get reassigned since controllerContext.partitionAssignments is an empty set (this will also result in orphan partitions).  
Again, I don't see the issue in this implementation, but I think it is worth checking. 

In addition, this doesn't address the case when there is controller move right after partition expansion (before the rearrange actually happened).  I think it is ok since we are not trying to handle 100% no replicas in the preferred controller.  Maybe we can add some comments for this unhandled situations.

Reply:
Thanks for the comments. It's true that they are all newPartitions given controllerContext.partitionAssignments is an empty set, but they shouldn't have a replica on the noNewPartitionBrokers, thus they shouldn't be rearranged.

It's a good point that the current implementation cannot handle controller switches. Given it's safe to scan all topics' partitions on controller switch, I think it should be done during a controller switch. Thoughts?

Reply:
manual assignment for existing topics can still assign partitions to noNewPartitionBrokers.  I think we cannot guarantee that noNewPartitionBrokers won't get replicas. In addition, there are some small-time window that new replicas can still get assigned to preferred controllers due to the fact that preferred controller znode is emphermal znode (see the design doc for more detail). 

I think a safer way is to rely on the existence of partition state znode when performing rearrangement. 

scan all topics' partitions on controller switch =>  If we cannot guarantee 100% no replica in the preferred controllers, I think it is ok to not performing special handing during controller switch given the overhead/additional hacky code needed (up to you to make a decision).

Reply:
Upon closer look, I find that there is already logic to handle the controller switch over by calling the `rearrangePartitionReplicaAssignmentForNewPartitions` method inside `initializeControllerContext`.

Regarding the preferred controller znodes being ephemeral, it's kinda an orthogonal design issue that we could address independently.